### ISSUE TITLE
Create JIRA ticket for Problems

### ISSUE THREAD

**Issue by ghost** [keptn-sandbox/jira-service#5]
Enhancement Request
Extend the functionality of the service to create JIRA tickets based on keptn problem events.

This functionality will need a flag so that this functionality can be disabled.

**Comment by johannes-b** [comment]
# Event flow for Dynatrace problem to make Jira ticket

**Use Case:**
This is for the scenario to have Keptn listen for a Dynatrace problem to automatically create a Jira ticket.

**Definition of Done:**
- [ ] Create a Jira ticket when there is a `sh.keptn.event.problem.open` event with status: `OPEN`. 
- [ ] Update the Jira ticket when the problem goes from `OPEN` to `CLOSED`. 
- [ ] A flag for the Keptn jira-service should be available to disable this feature because the Tempest JIRA plugin will already be doing this if it is installed

---

Use https://sequencediagram.org/ to open the event stream:

---

```
title Event flow to make Jira ticket for a Dynatrace problem
 
participant Dynatrace

Dynatrace->keptn-api: Problem notification in DT sends: sh.keptn.events.problem

keptn-api->event-broker: sh.keptn.events.problem

activate dynatrace-service

event-broker-->dynatrace-service: sh.keptn.events.problem

dynatrace-service->event-broker: sh.keptn.event.problem.open
  
deactivate dynatrace-service

dynatrace-service-->Dynatrace: push eval done event

activate jira-service

event-broker-->jira-service: sh.keptn.event.problem.open 

note over jira-service: State=OPEN
 
jira-service->Jira: add new ticket to Jira
 
note over jira-service: State=CLOSED
 
jira-service->Jira: add comment to ticket about status change
 
jira-service-->Dynatrace: add Dynatrace problem comment with Jira URL and Jira status
 
deactivate jira-service

```

**Comment by ghost** [comment]
I have the create ticket logic coded. This is working using BasicAuth to the JIRA servers.

Updating a ticket on a `CLOSED` problem assumes that the PID of a `CLOSED` problem is identical to the `OPEN` PID.

That PID is sent into JIRA in the ticket body, so my method was:
1. Grab the incoming PID of the `CLOSED` event.
2. Use that PID to look up the JIRA ticket ID where the `text` contains the PID.
3. Push a comment to that ticket (and potentially close the issue).

The blocker is my unfamiliarity with OAuth.

However the JIRA search functionality (to lookup the ticket via PID) requires OAuth. I have no idea why they've chosen to require different auth mechanisms for different operations.

So I can complete DoD items 1 and 3 fairly soon, but 2 will require someone with more OAuth knowledge to code the client.

For reference, the search is:

```
POST https://abc123.atlassian.net/rest/api/2/search
BODY
{"jql":"project = <PROJECT-KEY-FROM-SECRET> AND text ~ <PID-HERE>","startAt":0,"maxResults":1,"fields":["id","key"]}
```

**Comment by ghost** [comment]
1 and 3 are done and available in #4

**Comment by ghost** [comment]
Working notes for DoD 2.

### Problem Close Event
Problems are closed using the `sh.keptn.events.problem` with `payload.data.State = "RESOLVED"`

```
{
  "contenttype": "application/json",
  "data": {
    "ImpactedEntities": [
      {
        "entity": "SERVICE-A46EF51A4526E55C",
        "name": "unleash-demo:80",
        "type": "SERVICE"
      }
    ],
    "ImpactedEntity": "Failure rate increase on Web request service unleash-demo:80",
    "PID": "524042482955511204",
    "ProblemDetails": {
      "affectedCounts": {
        "APPLICATION": 0,
        "ENVIRONMENT": 0,
        "INFRASTRUCTURE": 0,
        "SERVICE": 0
      },
      "commentCount": 4,
      "displayName": "204",
      "endTime": 1588742973078,
      "hasRootCause": false,
      "id": "524042482955511204_1588742220000V2",
      "impactLevel": "SERVICE",
      "rankedEvents": [
        {
          "affectedRequestsPerMinute": 5.2,
          "endTime": 1588742973078,
          "entityId": "SERVICE-A46EF51A4526E55C",
          "entityName": "unleash-demo:80",
          "eventType": "FAILURE_RATE_INCREASED",
          "impactLevel": "SERVICE",
          "isRootCause": false,
          "service": "unleash-demo:80",
          "serviceMethodGroup": "Dynamic web requests",
          "severities": [
            {
              "context": "FAILURE_RATE",
              "unit": "Ratio",
              "value": 1
            }
          ],
          "severityLevel": "ERROR",
          "startTime": 1588742220000,
          "status": "CLOSED",
          "userDefinedFailureRateThreshold": 0
        }
      ],
      "rankedImpacts": [
        {
          "entityId": "SERVICE-A46EF51A4526E55C",
          "entityName": "unleash-demo:80",
          "eventType": "FAILURE_RATE_INCREASED",
          "impactLevel": "SERVICE",
          "severityLevel": "ERROR"
        }
      ],
      "recoveredCounts": {
        "APPLICATION": 0,
        "ENVIRONMENT": 0,
        "INFRASTRUCTURE": 0,
        "SERVICE": 1
      },
      "severityLevel": "ERROR",
      "startTime": 1588742220000,
      "status": "CLOSED",
      "tagsOfAffectedEntities": [
        {
          "context": "CONTEXTLESS",
          "key": "keptn_service",
          "value": "front-end"
        },
        {
          "context": "CONTEXTLESS",
          "key": "keptn_stage",
          "value": "production"
        },
        {
          "context": "CONTEXTLESS",
          "key": "keptn_project",
          "value": "website"
        }
      ]
    },
    "ProblemID": "204",
    "ProblemTitle": "Failure rate increase",
    "State": "RESOLVED",
    "Tags": "keptn_project:website, keptn_stage:production, keptn_service:front-end",
    "eventContext": null
  },
  "id": "197fc786-2731-409a-aa9b-7a3cf481fb67",
  "source": "https://github.com/keptn/keptn/api",
  "specversion": "0.2",
  "time": "2020-05-06T05:30:49.064Z",
  "type": "sh.keptn.events.problem",
  "shkeptncontext": "35323430-3432-4438-b239-353535313132",
  "label": "Problem resolved"
}
```

### Search for Issue ID using PID
```
POST https://abc123.atlassian.net/rest/api/2/search
{"jql":"project = PROJ AND text ~ 'PID: 1234'","startAt":0,"maxResults":1,"fields":["id","key"]}
```
Returns:
```
{
    "expand": "names,schema",
    "startAt": 0,
    "maxResults": 1,
    "total": 1,
    "issues": [
        {
            "expand": "operations,versionedRepresentations,editmeta,changelog,renderedFields",
            "id": "10240",
            "self": "https://abc123.atlassian.net/rest/api/2/issue/10240",
            "key": "PROJ-208"
        }
    ]
}
```

### Get Transitions for Issue
```
GET https://abc123.atlassian.net/rest/api/2/issue/PROJ-208/transitions
```
Returns:
```
{
    "expand": "transitions",
    "transitions": [
        {
            "id": "11",
            "name": "To Do",
            "to": {
                "self": "https://abc123.atlassian.net/rest/api/2/status/10000",
                "description": "",
                "iconUrl": "https://abc123.atlassian.net/",
                "name": "To Do",
                "id": "10000",
                "statusCategory": {
                    "self": "https://abc123.atlassian.net/rest/api/2/statuscategory/2",
                    "id": 2,
                    "key": "new",
                    "colorName": "blue-gray",
                    "name": "To Do"
                }
            },
            "hasScreen": false,
            "isGlobal": true,
            "isInitial": false,
            "isAvailable": true,
            "isConditional": false
        },
        ...
        ...
        {
            "id": "41",
            "name": "Done",
            "to": {
                "self": "https://abc123.atlassian.net/rest/api/2/status/10002",
                "description": "",
                "iconUrl": "https://abc123.atlassian.net/",
                "name": "Done",
                "id": "10002",
                "statusCategory": {
                    "self": "https://abc123.atlassian.net/rest/api/2/statuscategory/3",
                    "id": 3,
                    "key": "done",
                    "colorName": "green",
                    "name": "Done"
                }
            },
            "hasScreen": false,
            "isGlobal": true,
            "isInitial": false,
            "isAvailable": true,
            "isConditional": false
        }
    ]
}
```

### Post Comment to Issue

```
PUT https://abc123.atlassian.net/rest/api/2/issue/PROJ-208
{
   "update": {
      "comment": [{
            "add": {
               "body": "new comment"
            }
         }]
   }
}

```

### Transition The Issue
Use the `id` field. For example, to transition to `Done`:

```
POST https://abc123.atlassian.net/rest/api/2/issue/PROJ-208/transitions
{
   "transition": {"id": "41"}
}
```

@robertjahn Does the TEMPEST plugin close JIRA tickets for closed DT problems? If so, this service shouldn't interfere with that logic.
